scalafmt-sbt/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala

```diff
@@ -1,25 +1,194 @@
+import sbt.util.Logger
+
 import scala.meta.internal.tokenizers.PlatformTokenizerCache
-import sbt._
-import Keys._
+import scala.util.{Failure, Success, Try}
 
 object ScalafmtPlugin extends AutoPlugin {
   override def trigger: PluginTrigger = allRequirements
+
   object autoImport {
-    val scalafmt: Command =
-      Command.args("scalafmt", "run the scalafmt command line interface.") {
-        case (state, args) =>
-          org.scalafmt.cli.Cli.main("--non-interactive" +: args.toArray)
-          PlatformTokenizerCache.megaCache.clear()
-          state
+    val scalafmt = taskKey[Unit]("Format Scala sources.")
+    val scalafmtCheck =
+      taskKey[Boolean](
+        "Fails if a Scala source is mis-formatted. Does not write to files.")
+    val scalafmtOnCompile =
+      settingKey[Boolean](
+        "Format Scala source files on compile, off by default.")
+    val scalafmtConfig = taskKey[Option[File]](
+      "Optional location of .scalafmt.conf file. If None the default config is used.")
+    val scalafmtSbt = taskKey[Unit](
+      "Format *.sbt and project/*.scala files for this sbt build.")
+    val scalafmtSbtCheck =
+      taskKey[Boolean](
+        "Fails if a *.sbt or project/*.scala source is mis-formatted. Does not write to files.")
+    val scalafmtOnly = inputKey[Unit]("Format only given file.")
+  }
+  import autoImport._
+
+  private val scalafmtDoFormatOnCompile =
+    taskKey[Unit]("Format Scala source files if scalafmtOnCompile is on.")
+
+  private val scalaConfig =
+    scalafmtConfig.map(_.map(Config.fromHoconFile(_) match {
+      case Configured.Ok(conf) => conf
+      case Configured.NotOk(e) => throw new MessageOnlyException(e.msg)
+    }).getOrElse(ScalafmtConfig.default))
+  private val sbtConfig = scalaConfig.map(_.forSbt)
+
+  private type Input = String
+  private type Output = String
+
+  private def withFormattedSources[T](
+      sources: Seq[File],
+      config: ScalafmtConfig
+  )(
+      onError: (File, Throwable) => T,
+      onFormat: (File, Input, Output) => T
+  ): Seq[Option[T]] = {
+    sources.map(
+      file => {
+        val input = IO.read(file)
+        val output = Scalafmt.format(input, config)
+
+        output match {
+          case Formatted.Failure(e) =>
+            if (config.runner.fatalWarnings) {
+              throw e
+            } else if (config.runner.ignoreWarnings) {
+              // do nothing
+              None
+            } else {
+              Some(onError(file, e))
+            }
+          case Formatted.Success(code) =>
+            Some(onFormat(file, input, code))
+        }
+      }
+    )
+  }
+
+  private def formatSources(
+      sources: Seq[File],
+      config: ScalafmtConfig,
+      log: Logger
+  ): Unit = {
+    withFormattedSources(sources, config)(
+      (file, e) => log.error(s"Error in ${file.toString}: $e"),
+      (file, input, output) => {
+        if (input != output) {
+          log.info(s"Successfully formatted ${file.toString}.")
```

Comment:
This log entry will be quite noisy for large projects with 100s of sources. I propose we call this once per `formatSources` with a message like "Formatting 1 Scala source"/"Formatting 2 Scala sources" similar to how scalariform does https://github.com/sbt/sbt-scalariform/blob/a378b724a76ee94819939efc0e9adebd896e57d9/src/main/scala/com/typesafe/sbt/Scalariform.scala#L80

Reply:
Output from testing plugin on akka build https://gist.github.com/olafurpg/7a33d069604324a8bf4249cb067b7dd0